File: backend/app/schemas/book.py
from __future__ import annotations
from datetime import datetime
from typing import Literal, Optional, List
from pydantic import BaseModel, field_validator
from .category import CategoryBrief
from .location import LocationBrief
from .tag import TagResponse


BookStatus = Literal["available", "borrowed", "lost", "pending", "gifted"]
ReadStatus = Literal["unread", "reading", "read", "paused"]
BookSource = Literal["manual", "isbn_lookup", "title_search", "ollama", "import"]


class BookBase(BaseModel):
    title: str
    subtitle: Optional[str] = None
    author: Optional[str] = None
    translator: Optional[str] = None
    publisher: Optional[str] = None
    publish_year: Optional[int] = None
    isbn: Optional[str] = None
    original_isbn: Optional[str] = None
    language: Optional[str] = None
    pages: Optional[int] = None
    price_cents: Optional[int] = None
    binding: Optional[str] = None
    series: Optional[str] = None
    cover_url: Optional[str] = None
    summary: Optional[str] = None
    author_intro: Optional[str] = None
    category_id: Optional[int] = None
    location_id: Optional[int] = None
    status: BookStatus = "available"
    read_status: ReadStatus = "unread"
    rating: Optional[int] = None
    is_favorite: bool = False
    note: Optional[str] = None

    @field_validator("rating")
    @classmethod
    def rating_range(cls, v: Optional[int]) -> Optional[int]:
        if v is not None and not (1 <= v <= 5):
            raise ValueError("评分必须在 1-5 之间")
        return v


class BookCreate(BookBase):
    source: BookSource = "manual"
    tag_names: List[str] = []


class BookUpdate(BaseModel):
    title: Optional[str] = None
    subtitle: Optional[str] = None
    author: Optional[str] = None
    translator: Optional[str] = None
    publisher: Optional[str] = None
    publish_year: Optional[int] = None
    isbn: Optional[str] = None
    language: Optional[str] = None
    pages: Optional[int] = None
    price_cents: Optional[int] = None
    binding: Optional[str] = None
    series: Optional[str] = None
    cover_url: Optional[str] = None
    summary: Optional[str] = None
    author_intro: Optional[str] = None
    category_id: Optional[int] = None
    location_id: Optional[int] = None
    status: Optional[BookStatus] = None
    read_status: Optional[ReadStatus] = None
    rating: Optional[int] = None
    is_favorite: Optional[bool] = None
    note: Optional[str] = None
    tag_names: Optional[List[str]] = None

    @field_validator("rating")
    @classmethod
    def rating_range(cls, v: Optional[int]) -> Optional[int]:
        if v is not None and not (1 <= v <= 5):
            raise ValueError("评分必须在 1-5 之间")
        return v


class BookListItem(BaseModel):
    id: int
    title: str
    subtitle: Optional[str] = None
    author: Optional[str] = None
    publisher: Optional[str] = None
    publish_year: Optional[int] = None
    isbn: Optional[str] = None
    cover_url: Optional[str] = None
    category: Optional[CategoryBrief] = None
    location: Optional[LocationBrief] = None
    status: BookStatus
    read_status: ReadStatus
    rating: Optional[int] = None
    is_favorite: bool
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class BookResponse(BookBase):
    id: int
    source: BookSource
    category: Optional[CategoryBrief] = None
    location: Optional[LocationBrief] = None
    tags: List[TagResponse] = []
    created_by: Optional[int] = None
    updated_by: Optional[int] = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class BookBatchUpdate(BaseModel):
    book_ids: List[int]
    updates: dict


class PaginatedBooks(BaseModel):
    items: List[BookListItem]
    total: int
    page: int
    page_size: int

File: frontend/src/types/book.ts
export type BookStatus = 'available' | 'borrowed' | 'lost' | 'pending' | 'gifted';
export type ReadStatus = 'unread' | 'reading' | 'read' | 'paused';
export type BookSource = 'manual' | 'isbn_lookup' | 'title_search' | 'ollama' | 'import';

export interface CategoryOption {
  id: number;
  code: string;
  name: string;
  parentId: number | null;
  description?: string;
  sortOrder?: number;
  isSystem?: boolean;
  children?: CategoryOption[];
}

export interface LocationOption {
  id: number;
  room: string;
  shelf: string;
  layer?: string;
  position?: string;
  fullPath: string;
  description?: string;
  sortOrder?: number;
}

export interface BookTag {
  id?: number;
  name: string;
  color?: string;
}

export interface BookListItem {
  id: number;
  title: string;
  subtitle: string;
  author: string;
  publisher: string;
  publishYear: number | null;
  isbn: string;
  coverUrl: string;
  category: CategoryOption | null;
  location: Pick<LocationOption, 'id' | 'fullPath'> | null;
  status: BookStatus;
  readStatus: ReadStatus;
  rating: number | null;
  isFavorite: boolean;
  createdAt: string;
  updatedAt: string;
}

export interface BookDetail extends BookListItem {
  translator: string;
  originalIsbn: string;
  language: string;
  pages: number | null;
  priceCents: number | null;
  binding: string;
  series: string;
  summary: string;
  authorIntro: string;
  tags: BookTag[];
  source: BookSource;
  note: string;
  categoryId: number | null;
  locationId: number | null;
}

export interface BookQuery {
  page: number;
  pageSize: number;
  keyword?: string;
  categoryId?: number;
  locationId?: number;
  status?: BookStatus;
  readStatus?: ReadStatus;
  isFavorite?: boolean;
}

export interface PaginatedBooks {
  items: BookListItem[];
  total: number;
  page: number;
  pageSize: number;
}

export interface BookFormModel {
  title: string;
  subtitle: string;
  author: string;
  translator: string;
  publisher: string;
  publishYear: number | null;
  isbn: string;
  language: string;
  pages: number | null;
  priceCents: number | null;
  binding: string;
  series: string;
  coverUrl: string;
  summary: string;
  authorIntro: string;
  categoryId: number | null;
  locationId: number | null;
  tagNames: string[];
  status: BookStatus;
  readStatus: ReadStatus;
  rating: number | null;
  isFavorite: boolean;
  note: string;
}

export interface BookBatchUpdatePayload {
  bookIds: number[];
  updates: Partial<Pick<BookFormModel, 'categoryId' | 'locationId' | 'status' | 'readStatus' | 'tagNames'>>;
}

export const BOOK_STATUS_OPTIONS: Array<{ label: string; value: BookStatus }> = [
  { label: '在架', value: 'available' },
  { label: '借出', value: 'borrowed' },
  { label: '遗失', value: 'lost' },
  { label: '待整理', value: 'pending' },
  { label: '已转赠', value: 'gifted' },
];

export const READ_STATUS_OPTIONS: Array<{ label: string; value: ReadStatus }> = [
  { label: '未读', value: 'unread' },
  { label: '阅读中', value: 'reading' },
  { label: '已读', value: 'read' },
  { label: '暂停', value: 'paused' },
];

export function createEmptyBookForm(): BookFormModel {
  return {
    title: '',
    subtitle: '',
    author: '',
    translator: '',
    publisher: '',
    publishYear: null,
    isbn: '',
    language: 'zh-CN',
    pages: null,
    priceCents: null,
    binding: '',
    series: '',
    coverUrl: '',
    summary: '',
    authorIntro: '',
    categoryId: null,
    locationId: null,
    tagNames: [],
    status: 'available',
    readStatus: 'unread',
    rating: null,
    isFavorite: false,
    note: '',
  };
}
